base/sparse/sparsematrix.jl

```diff
@@ -2252,6 +2252,8 @@ function setindex!{Tv}(A::SparseMatrixCSC{Tv}, x::Number,
     if isempty(I) || isempty(J); return A; end
     if !issorted(I); I = sort(I); end
     if !issorted(J); J = sort(J); end
+    if !allunique(I); I = unique(I); end
```

Comment:
I wasn't seeing the whole function, but you're right, here `I` or `J` shouldn't be mutated. Another option that comes to mind then is

```julia
if !issorted(I, lt=≤); I = !allunique(I) ? sort!(unique(I)) : sort(I); end
```